File: poh5/poh5_4b.py
#!/usr/bin/python
import sys
if sys.version_info[0]>=3: raw_input=input

W,H,Q=[int(e) for e in raw_input().split()]
m=[[int(e) for e in raw_input().split()] for i in range(H)]
imos=[[0]*(W+1) for i in range(H+1)]
for i in range(Q):
	w1,h1,w2,h2=[int(e) for e in raw_input().split()]
	imos[h1-1][w1-1]+=1
	imos[h1-1][w2]-=1
	imos[h2][w1-1]-=1
	imos[h2][w2]+=1
s=0
for h in range(0,H):
	for w in range(1,W):
		imos[h][w]+=imos[h][w-1]
for h in range(1,H):
	for w in range(0,W):
		imos[h][w]+=imos[h-1][w]
for h in range(0,H):
	for w in range(0,W):
		if imos[h][w]: s+=m[h][w]
print(s)

File: poh5/poh5_4b.rb
#!/usr/bin/ruby
W,H,Q=gets.split.map(&:to_i)
m=H.times.map{gets.split.map(&:to_i)}
imos=(H+1).times.map{[0]*(W+1)}
Q.times{
	w1,h1,w2,h2=gets.split.map(&:to_i)
	imos[h1-1][w1-1]+=1
	imos[h1-1][w2]-=1
	imos[h2][w1-1]-=1
	imos[h2][w2]+=1
}
s=0
0.upto(H-1){|h|1.upto(W-1){|w|imos[h][w]+=imos[h][w-1]}}
1.upto(H-1){|h|0.upto(W-1){|w|imos[h][w]+=imos[h-1][w]}}
0.upto(H-1){|h|0.upto(W-1){|w|s+=m[h][w] if imos[h][w]>0}}
p s